internal/pginfer/pginfer.go

```diff
@@ -143,75 +239,117 @@ func (inf *Inferrer) prepareTypes(query *ast.SourceQuery) (_a []InputParam, _ []
 			// only if there's output columns. If the user has an UPDATE or INSERT
 			// without a RETURNING clause, pggen will surface the null constraint
 			// errors because len(descriptions) == 0.
-			if strings.Contains(strings.ToLower(query.PreparedSQL), "update") ||
-				strings.Contains(strings.ToLower(query.PreparedSQL), "insert") {
+			if strings.Contains(strings.ToLower(sql), "update") ||
+				strings.Contains(strings.ToLower(sql), "insert") {
 				msg += "\n    HINT: if the main statement is an UPDATE or INSERT ensure that you have"
-				msg += "\n          a RETURNING clause (this query is marked " + string(query.ResultKind) + ")."
+				msg += "\n          a RETURNING clause (this query is marked " + string(resultKind) + ")."
 				msg += "\n          Use :exec if you don't need the query output."
 			}
-			return nil, nil, fmt.Errorf(msg+"\n    %w", pgErr)
+			return nil, fmt.Errorf(msg+"\n    %w", pgErr)
 		}
-		return nil, nil, fmt.Errorf("prepare query to infer types: %w", err)
+		return nil, fmt.Errorf("prepare query to infer types: %w", err)
 	}
 
-	// Validate.
-	if len(stmtDesc.ParamOIDs) != len(query.Params) {
-		return nil, nil, fmt.Errorf("expected %d parameter types for query; got %d", len(query.Params), len(stmtDesc.ParamOIDs))
+	return stmtDesc, nil
+}
+
+func (inf *Inferrer) getInputParamTypes(statements []*pgconn.StatementDescription, query *ast.SourceQuery) ([]InputParam, error) {
+	var oids []uint32
+	oidToParamIndex := make(map[uint32][]int)
+
+	setParams := make([]bool, len(query.Params))
+
+	if len(statements) != len(query.DenseSQL) {
+		return nil, fmt.Errorf("expected %d statement descriptions; got %d", len(query.DenseSQL), len(statements))
 	}
 
-	// Build input params.
-	var inputParams []InputParam
-	if len(stmtDesc.ParamOIDs) > 0 {
-		types, err := inf.typeFetcher.FindTypesByOIDs(stmtDesc.ParamOIDs...)
-		if err != nil {
-			return nil, nil, fmt.Errorf("fetch oid types: %w", err)
+	for i, statement := range statements {
```

Comment:
Everything is complicated by having multiple statements but don't let the messy diff in this function fool you, this is ultimately mostly a refactor to add a for loop and an extra parameter index remapping. A lot of the logic already existed.